### PR TITLE
Initial attempt at merging the work on the TimeFrame done so far

### PR DESCRIPTION
**Pull request by ktf**
Do not merge. This is just to allow people (e.g. @matthiasrichter ) to comment on the progress so far.

Things to fix:

- [x]  Please disentangle commit ac50210 from this PR; merge as a separate PR and rebase the source branch.
- [x] Commit f40eb44 must be removed; it wrongly reverts changes of commit 1f1700b. In fact the initial changes of this commit have been merged already in PR #241; now it is a wrong duplicate in the history
- [x] Drop `set(CMAKE_EXPORT_COMPILE_COMMANDS 1)`
- [x] Move `SubframeMetadata.h` to `DataFormats/Headers`
- [x] Revert dependencies for examples to their original ones.
- [x] Add `...Device...` to `EPNReceiver.*` and `FLPSender.*` filenames.
- [x] Squash in less / one commits.

### CODE REVIEW

**Comment by sawenzel** [comment]
@ktf: This PR is missing a few things that we had before (e.g., the python script to scale the setup, etc.)

File: DataFormats/Headers/test/dataHeaderTest.cxx [hunk 1]
```diff
@@ -7,51 +7,6 @@
 namespace AliceO2 {
   namespace Header {
 
-    BOOST_AUTO_TEST_CASE(Description_test)
```
Comment (matthiasrichter) [inline DataFormats/Headers/test/dataHeaderTest.cxx]: that is surely a mistake, removes all changes made in 1f1700b7fdbd2a30ede283ddd9a9cc9693d3e5f9. This is definitely an incorrectly resolved merge conflict.

As it looks now, commit f40eb44aa31b5e97b373044d4bdc9b49a6800a54 must be removed. Actually Sandro's change was added in #241.

File: Utilities/DataFlow/include/DataFlow/SubframeMetadata.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#ifndef SUBFRAMEMETADATA_H
```
Comment (matthiasrichter) [inline Utilities/DataFlow/include/DataFlow/SubframeMetadata.h]: move to `DataFormat/Headers/include/Headers`

File: Utilities/DataFlow/run/CreateSetup.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/python
```
Comment (matthiasrichter) [inline Utilities/DataFlow/run/CreateSetup.py]: fantastic!

File: Utilities/DataFlow/run/CreateSetup.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+# a very simple python script to
+# create and scale a topology/conf + runfile for a variable list of
+# HeartbeatSampler + DataPublisher + ... + FLP + EPN
+# author: S. Wenzel
+#FIXME: decide whether to config for xterm or tmux
```
Comment (matthiasrichter) [inline Utilities/DataFlow/run/CreateSetup.py]: just add a few lines of instructions how to use it

File: Utilities/DataFlow/run/confComplexSetup.json [hunk 5]
```diff
@@ -0,0 +1,132 @@
+{
```
Comment (matthiasrichter) [inline Utilities/DataFlow/run/confComplexSetup.json]: does the python script make this obsolete?

File: cmake/O2Dependencies.cmake [hunk 6]
```diff
@@ -136,6 +136,7 @@ o2_define_bucket(
     DEPENDENCIES
     Base
     Headers
+    DataFlow
```
Comment (matthiasrichter) [inline cmake/O2Dependencies.cmake]: not sure about this, comes probably from the fact that SubframeMetadata.h is in the wrong place

File: cmake/O2Dependencies.cmake [hunk 7]
```diff
@@ -163,6 +164,7 @@ o2_define_bucket(
     ${OPTIONAL_DDS_LIBRARIES}
     Base
     Headers
+    DataFlow
```
Comment (matthiasrichter) [inline cmake/O2Dependencies.cmake]: can be removed, we decided to leave the FLP-EPNdistributed example unchanged

File: CMakeLists.txt [hunk 8]
```diff
@@ -8,6 +8,8 @@
 # Check IF cmake has the required version
 CMAKE_MINIMUM_REQUIRED(VERSION 2.8.11 FATAL_ERROR)
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS 1)
```
Comment (sawenzel) [inline CMakeLists.txt]: It's not. This is for better tool support (automatic C++ completion in editors) + static code analysis. It should be merged as a separate PR.

File: Utilities/DataFlow/run/CreateSetup.py (hunk 4 above)
Comment (sawenzel) [inline Utilities/DataFlow/run/CreateSetup.py]: Will do.

File: Utilities/DataFlow/run/confComplexSetup.json (hunk 5 above)
Comment (sawenzel) [inline Utilities/DataFlow/run/confComplexSetup.json]: in principle yes

File: CMakeLists.txt (hunk 8 above)
Comment (matthiasrichter) [inline CMakeLists.txt]: ah, great! Yes, please make this a separate PR

File: Utilities/DataFlow/include/DataFlow/EPNReceiver.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
+/**
```
Comment (matthiasrichter) [inline Utilities/DataFlow/include/DataFlow/EPNReceiver.h]: at some point we came to the conclusion that devices should be named `***Device`, at least for the new devices we should stick to this.

File: CMakeLists.txt (hunk 8 above)
Comment (ktf) [inline CMakeLists.txt]: I would actually drop this and leave it to the user. Notice that if you use alibuild this will nevertheless set to TRUE.

**Comment by ktf** [comment]
I updated the pull request description with the list of action items (so that we do not lose  them) and will soon proceed with the rebase.

**Comment by ktf** [comment]
Ok, I've taken care of all your comments. Tomorrow we can sit together and squash things here and there and then we reiterate.

**Comment by ktf** [comment]
Ok, I've squashed the work in a few commits, feel free to give an additional look. Some personal comments:

- [ ] `SubtimeFrameMetadata` has a few dummy classes for test data. Maybe we should actually split them off in a separate header (`DataFormats/DummyData.h`?)
- [ ] There is still opportunity for a few clean ups (e.g. we have commented code here and there).
- [ ] HeartbeatSampler can probably be renamed to HeartbeatSamplerDevice

Do we agree?

**Comment by matthiasrichter** [comment]
Thanks a lot, please be a bit patient, I won't manage today to look into the code.

File: Utilities/DataFlow/include/DataFlow/HeartbeatSampler.h [hunk 10]
```diff
@@ -31,14 +31,11 @@ class HeartbeatSampler : public Base::O2Device
 public:
   typedef AliceO2::Base::O2Message O2Message;
 
-  static constexpr const char* OptionKeyOutputChannelName = "output-channel-name";
+  static constexpr const char* OptionKeyOutputChannelName = "out-chan-name";
```
Comment (matthiasrichter) [inline Utilities/DataFlow/include/DataFlow/HeartbeatSampler.h]: Probably I forgot to comment on that last time, I thought I had spotted this. What's the reason for the renaming? For me `--output-channel-name` is the better keyword. But that's a minor detail.

**Comment by ktf** [comment]
This was done because everything else uses "out-chan-name" and that was
confusing. I propose we keep it as in the PR, and then we bulk rename all
of them to a better name.
